### PR TITLE
Fix __getitem__() for various index types

### LINKED ISSUE #536: PeriodIndex element `start_time` and `end_time` attributes raise error
**Describe the bug**
Accessing `start_time` or `end_time` on a `PeriodIndex` raises a `Member <> is unknown` error.

**To Reproduce**
Code:
```python3
import pandas as pd
index = pd.period_range(start = "2000", end="2023", freq="y")
index[0].start_time
```
Error:
```
/home/l05/code/opensource/reproduce/pandas_start_end_time.py
  /home/l05/code/opensource/reproduce/pandas_start_end_time.py:4:10 - error: Cannot access member "start_time" for type "str"
    Member "start_time" is unknown (reportGeneralTypeIssues)
  /home/l05/code/opensource/reproduce/pandas_start_end_time.py:4:10 - error: Cannot access member "start_time" for type "bytes"
    Member "start_time" is unknown (reportGeneralTypeIssues)
  /home/l05/code/opensource/reproduce/pandas_start_end_time.py:4:10 - error: Cannot access member "start_time" for type "date"
    Member "start_time" is unknown (reportGeneralTypeIssues)
  /home/l05/code/opensource/reproduce/pandas_start_end_time.py:4:10 - error: Cannot access member "start_time" for type "datetime"
    Member "start_time" is unknown (reportGeneralTypeIssues)
  /home/l05/code/opensource/reproduce/pandas_start_end_time.py:4:10 - error: Cannot access member "start_time" for type "timedelta"
    Member "start_time" is unknown (reportGeneralTypeIssues)
  /home/l05/code/opensource/reproduce/pandas_start_end_time.py:4:10 - error: Cannot access member "start_time" for type "bool"
    Member "start_time" is unknown (reportGeneralTypeIssues)
  /home/l05/code/opensource/reproduce/pandas_start_end_time.py:4:10 - error: Cannot access member "start_time" for type "int"
    Member "start_time" is unknown (reportGeneralTypeIssues)
  /home/l05/code/opensource/reproduce/pandas_start_end_time.py:4:10 - error: Cannot access member "start_time" for type "float"
    Member "start_time" is unknown (reportGeneralTypeIssues)
  /home/l05/code/opensource/reproduce/pandas_start_end_time.py:4:10 - error: Cannot access member "start_time" for type "Timestamp"
    Member "start_time" is unknown (reportGeneralTypeIssues)
  /home/l05/code/opensource/reproduce/pandas_start_end_time.py:4:10 - error: Cannot access member "start_time" for type "Timedelta"
    Member "start_time" is unknown (reportGeneralTypeIssues)
  /home/l05/code/opensource/reproduce/pandas_start_end_time.py:4:10 - error: Cannot access member "start_time" for type "complex"
    Member "start_time" is unknown (reportGeneralTypeIssues)
```

**Please complete the following information:**
 - OS: Ubuntu 20.04
 - python version: 3.10.7
 - version of type checker: pyright 1.1.293
 - version of installed `pandas-stubs`: 1.5.3.230203

**Additional context**
Would this be as simple as updating the [`IndexIterScaler`](https://github.com/pandas-dev/pandas-stubs/blob/c5d9c0802d2e07d9b4e4ca06fba71928cd720c19/pandas-stubs/_typing.pyi#L171) type? I would be happy to open a PR.

### PR DESCRIPTION
**Pull request by Dr-Irv**
- [x] Closes #536
- [x] Tests added: 
  - `test_indexes.py:test_getitem()`

By making `__getitem__()` a mixin, it uncovered some odd hierarchy in the index types, where we had redundant declarations, and some incorrect detection of some invalid subtraction operations.  So the changes here make the original tests pass, and then we can test `__getitem__()` on various index types.

Was able to get rid of a number of `#type: ignore` statements as a result.

Either @bashtage  or @twoertwein  can review/approve.

### CODE REVIEW

File: pandas-stubs/core/indexes/base.pyi [hunk 1]
```diff
@@ -48,6 +50,22 @@ class InvalidIndexError(Exception): ...
 
 _str = str
 
+class _IndexGetitemMixin(Generic[S1]):
+    # type ignore needed because it doesn't like the type of self
+    @overload
+    def __getitem__(  # type: ignore[misc]
+        self: IndexT,
```
Comment (twoertwein) [inline pandas-stubs/core/indexes/base.pyi:57]: I assume `only` Index inherits from this. Can you use `_IndexGetitemMixinT` for self/return value?

Comment (twoertwein) [reply pandas-stubs/core/indexes/base.pyi:57]: Could probably also start using `typing.Self`

Comment (Dr-Irv) [reply pandas-stubs/core/indexes/base.pyi]: > I assume `only` Index inherits from this. Can you use `_IndexGetitemMixinT` for self/return value?

Your assumption isn't correct.  `DatetimeIndex`, `PeriodIndex` and `TimedeltaIndex` also inherit, and then are specifying the return type of `__getitem__(x: int)` via the `Generic` parameter.  

> Could probably also start using typing.Self

I tried creating a `TypeVar` as  `_IndexGetitemMixinT` and it did work.  Tried `typing.Self`, and it doesn't work.  I think the type of `self` in this case has to be "bound" to the return type.

These changes are in commit https://github.com/pandas-dev/pandas-stubs/pull/537/commits/0db8a00a7bd5c7f023991459d79f372df99b1ab3

**Review by twoertwein — APPROVED** [review]

**Comment by twoertwein** [comment]
Thank you @Dr-Irv !

**Pull request merged**